File: JS/NumberOfZigzagArrays_I.js
var zigZagArrays = function(n, l, r) {
    const MOD = 1000000007;
    const m = r - l + 1;
    let up = new Array(m).fill(1);
    let down = new Array(m).fill(1);
    let pre = new Array(m + 1).fill(0);
    let suf = new Array(m + 1).fill(0);
    let newUp = new Array(m);
    let newDown = new Array(m);

    for (let len = 2; len <= n; len++) {
        pre[0] = 0;
        suf[m] = 0;
        for (let i = 0; i < m; i++) { pre[i + 1] = (pre[i] + down[i]) % MOD; }
        for (let i = m - 1; i >= 0; i--) { suf[i] = (suf[i + 1] + up[i]) % MOD; }
        for (let i = 0; i < m; i++) { newUp[i] = pre[i]; newDown[i] = suf[i + 1]; }
        [up, newUp] = [newUp, up];
        [down, newDown] = [newDown, down];
    }

    let ans = 0;
    for (let i = 0; i < m; i++) { ans = (ans + up[i] + down[i]) % MOD; }
    return ans;
};

File: Python/NumberOfZigzagArrays_I.py
class Solution(object):
    def zigZagArrays(self, n, l, r):
        MOD = 1000000007
        m = r - l + 1

        up = [1] * m
        down = [1] * m

        for _ in range(2, n + 1):
            pre = [0] * (m + 1)
            suf = [0] * (m + 1)

            for i in range(m):
                pre[i + 1] = (pre[i] + down[i]) % MOD

            for i in range(m - 1, -1, -1):
                suf[i] = (suf[i + 1] + up[i]) % MOD

            newUp = [0] * m
            newDown = [0] * m

            for i in range(m):
                newUp[i] = pre[i]
                newDown[i] = suf[i + 1]

            up = newUp
            down = newDown

        ans = 0
        for i in range(m):
            ans = (ans + up[i] + down[i]) % MOD

        return ans